File: src/day5.R
# Libraries ---------------------------------------------------------------
library(purrr)
library(stringr)
library(magrittr)
library(tidyr)
library(dplyr)

# Functions ---------------------------------------------------------------

# Convenience
x1 <- \(l) l[[1]][1]
x2 <- \(l) l[[2]][1]
y1 <- \(l) l[[1]][2]
y2 <- \(l) l[[2]][2]
px <- \(l) x1(l):x2(l)
py <- \(l) y1(l):y2(l)
dx <- \(l) x1(l) - x2(l)
dy <- \(l) y1(l) - y2(l)

# Expand a line into integer points
points <- function(line) {
  if (x1(line) == x2(line))
    map(py(line), ~c(x1(line), .))
  else if (y1(line) == y2(line))
    map(px(line), ~c(., y1(line)))
  else if (abs(dx(line)) == abs(dy(line)))
    map2(px(line), py(line), ~c(.x, .y))
  else
    stop("Must only be horizontal, vertical, or 45 diagonal")
}

# Get overlapping points between two lines
overlaps <- function(l1, l2) {
  intersect(points(l1), points(l2))
}

# Counting of overlaps
count_overlaps <- function(set) {
  combn(set, 2) %>%
    { tibble(line1 = .[1,], line2 = .[2,]) } %>%
    mutate(overlaps = map2(line1, line2, overlaps)) %>%
    filter(map_lgl(overlaps, ~length(.) > 0)) %>%
    pull(overlaps) %>%
    flatten() %>%
    n_distinct()
}

# Execute -----------------------------------------------------------------

# Parse the inputs
lines <-
  readLines("data/day5.txt") %>%
  map(str_split, " -> ") %>%
  map(extract2, 1) %>%
  map(function(line) {
    map(
      line,
      function(point) as.integer(str_split(point, ",")[[1]])
    )
  })

# Horizontal and Vertical Lines only
plines <- keep(lines, ~x1(.) == x2(.) | y1(.) == y2(.))

# Part 1 Answer: Number of Points with overlaps
part1_ans <- count_overlaps(plines)

# Part 2 Answer: Number of Points with overlaps (diagonals included)
part2_ans <- count_overlaps(lines)

File: src/day5.jl
module Day5

lines = map(readlines("data/day5.txt")) do line 
    p1, p2 = split(line, " -> ")
    x1, y1 = parse.(Int, split(p1, ","))
    x2, y2 = parse.(Int, split(p2, ",")) 
    return ((x1, y1), (x2, y2))
end

hvlines = filter(lines) do line
    x1, y1 = line[1]
    x2, y2 = line[2]
    return (x1 == x2) | (y1 == y2)
end

function points(line)
    x1, y1 = line[1]
    x2, y2 = line[2]
    px = x1:(x1 < x2 ? 1 : -1):x2
    py = y1:(y1 < y2 ? 1 : -1):y2
    if (x1 == x2) 
        return [(x1, y) for y in py]
    elseif (y1 == y2)
        return [(x, y1) for x in px]
    else 
        return [(x, y) for (x, y) in zip(px, py)]
    end
end

function count_overlaps(lines) 

    floor = zeros(1000, 1000)

    for path in map(points, lines)
        for point in path
            floor[point...] += 1
        end
    end

    return length(filter(x -> x >= 2, floor))

end

part1_ans = count_overlaps(hvlines)
part2_ans = count_overlaps(lines)

end